File: go/pkg/models/character.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type Character struct {
	ID                         uuid.UUID                   `json:"id" gorm:"type:uuid;default:uuid_generate_v4()"`
	CreatedAt                  time.Time                   `json:"createdAt"`
	UpdatedAt                  time.Time                   `json:"updatedAt"`
	Name                       string                      `json:"name"`
	Description                string                      `json:"description"`
	InternalTags               StringArray                 `json:"internalTags" gorm:"column:internal_tags;type:jsonb"`
	MapIconURL                 string                      `json:"mapIconUrl"`
	DialogueImageURL           string                      `json:"dialogueImageUrl"`
	ThumbnailURL               string                      `json:"thumbnailUrl"`
	OwnerUserID                *uuid.UUID                  `json:"ownerUserId,omitempty" gorm:"column:owner_user_id;type:uuid"`
	Ephemeral                  bool                        `json:"ephemeral" gorm:"column:ephemeral"`
	StoryVariants              CharacterStoryVariants      `json:"storyVariants" gorm:"column:story_variants;type:jsonb;default:'[]'"`
	Relationship               *CharacterRelationshipState `json:"relationship,omitempty" gorm:"-"`
	ImageGenerationStatus      string                      `json:"imageGenerationStatus" gorm:"column:image_generation_status"`
	ImageGenerationError       *string                     `json:"imageGenerationError,omitempty" gorm:"column:image_generation_error"`
	Locations                  []CharacterLocation         `json:"locations" gorm:"foreignKey:CharacterID"`
	PointOfInterestID          *uuid.UUID                  `json:"pointOfInterestId,omitempty" gorm:"type:uuid"`
	PointOfInterest            *PointOfInterest            `json:"pointOfInterest,omitempty" gorm:"foreignKey:PointOfInterestID"`
	HasAvailableQuest          bool                        `json:"hasAvailableQuest" gorm:"-"`
	HasAvailableMainStoryQuest bool                        `json:"hasAvailableMainStoryQuest" gorm:"-"`
}

const CharacterInternalTagGeneratedFetchQuest = "generated_fetch_quest_character"
const CharacterInternalTagGeneratedPOILocal = "generated_poi_local"

func CharacterHasInternalTag(character *Character, tag string) bool {
	if character == nil || tag == "" {
		return false
	}
	for _, existing := range character.InternalTags {
		if existing == tag {
			return true
		}
	}
	return false
}

func (n *Character) TableName() string {
	return "characters"
}

func (n *Character) BeforeSave(tx *gorm.DB) error {
	if n.InternalTags == nil {
		n.InternalTags = StringArray{}
	}
	if n.StoryVariants == nil {
		n.StoryVariants = CharacterStoryVariants{}
	}
	return nil
}

const (
	CharacterImageGenerationStatusNone       = "none"
	CharacterImageGenerationStatusQueued     = "queued"
	CharacterImageGenerationStatusInProgress = "in_progress"
	CharacterImageGenerationStatusComplete   = "complete"
	CharacterImageGenerationStatusFailed     = "failed"
)

File: dart/unclaimed_streets/lib/models/character.dart
import 'character_location.dart';

class Character {
  final String id;
  final String name;
  final String? description;
  final String? mapIconUrl;
  final String? dialogueImageUrl;
  final String? thumbnailUrl;
  final String? pointOfInterestId;
  final double? pointOfInterestLat;
  final double? pointOfInterestLng;
  final List<CharacterLocation> locations;
  final bool hasAvailableQuest;
  final bool hasAvailableMainStoryQuest;

  const Character({
    required this.id,
    required this.name,
    this.description,
    this.mapIconUrl,
    this.dialogueImageUrl,
    this.thumbnailUrl,
    this.pointOfInterestId,
    this.pointOfInterestLat,
    this.pointOfInterestLng,
    this.locations = const [],
    this.hasAvailableQuest = false,
    this.hasAvailableMainStoryQuest = false,
  });

  factory Character.fromJson(Map<String, dynamic> json) {
    double? parseCoordinate(dynamic raw) {
      if (raw == null) return null;
      if (raw is num) return raw.toDouble();
      if (raw is String) return double.tryParse(raw.trim());
      return null;
    }

    final poi = json['pointOfInterest'];
    double? pointOfInterestLat;
    double? pointOfInterestLng;
    if (poi is Map<String, dynamic>) {
      pointOfInterestLat = parseCoordinate(poi['lat'] ?? poi['latitude']);
      pointOfInterestLng = parseCoordinate(poi['lng'] ?? poi['longitude']);
    }

    return Character(
      id: json['id'] as String,
      name: json['name'] as String? ?? '',
      description: json['description'] as String?,
      mapIconUrl: json['mapIconUrl'] as String?,
      dialogueImageUrl: json['dialogueImageUrl'] as String?,
      thumbnailUrl: json['thumbnailUrl'] as String?,
      pointOfInterestId: json['pointOfInterestId'] as String?,
      pointOfInterestLat: pointOfInterestLat,
      pointOfInterestLng: pointOfInterestLng,
      locations:
          (json['locations'] as List<dynamic>?)
              ?.map(
                (e) => CharacterLocation.fromJson(e as Map<String, dynamic>),
              )
              .toList() ??
          const [],
      hasAvailableQuest: json['hasAvailableQuest'] as bool? ?? false,
      hasAvailableMainStoryQuest:
          json['hasAvailableMainStoryQuest'] as bool? ?? false,
    );
  }
}
